### PR TITLE
Add `Lint/ReadInstanceVar`

### LINKED ISSUE #463: Lint against accessing instance variables outside of the instance of a class
It's possible to access instance variables on an instance of a class similarly to `instance_variable_get` in Ruby

```crystal
class MyKlass
  def initialize(@my_instance_var : String); end
end

puts MyKlass.new("hello").@my_instance_var
# => "hello"
```
[Crystal Playground](https://play.crystal-lang.org/#/r/gzsz)

This can be handy for debugging purposes but I think generally should be discouraged as I think it's commonly expected that instance variables are private to the instance of a class.

### PR DESCRIPTION
**Pull request by nobodywasishere**
Closes #463

### CODE REVIEW

**Comment by straight-shoota** [comment]
IMHO such a rule is impossible to implement without taking scopes into account. Examples like the one in https://github.com/crystal-ameba/ameba/issues/463#issuecomment-2225517353 where an ivar is accessed from the same class should be fine.

Doing this correctly would require semantic analysis though.

**Comment by nobodywasishere** [comment]
IMO in that case I think the upstream issue should be closed as no-fix.

**Comment by Sija** [comment]
I think it should be doable with a low false-positive rate using a custom visitor, that ignores the access done for the object of the same class.

**Comment by nobodywasishere** [comment]
Hm okay, I'll try that. Do you mean something along the lines of class/struct ClassName > def.args.any? &.restriction == ClassName or self, then reading instance vars is allow to that variable?

**Comment by Sija** [comment]
Sth along these lines, yes. Top-level code as well needs to be taken into account.

File: src/ameba/rule/lint/read_instance_var.cr [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  # ```
+  # class Greeter
+  #   def combine(other)
+  #     @ivar <=> other.@ivar
+  #   end
+  #
+  #   def split(other : Smiler)
+  #     @ivar - other.@ivar
+  #   end
+  # end
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:19]: Let's make the examples more abstract (like below):

```suggestion
  # class Foo
  #   def <=>(other)
  #     @ivar <=> other.@ivar
  #   end
  #
  #   def <=>(other : Bar)
  #     @ivar <=> other.@ivar
  #   end
  # end
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  # ```
+  # class Greeter
+  #   def combine(other : self)
+  #     @ivar <=> other.@ivar
+  #   end
+  #
+  #   def split(other : Greeter)
+  #     @ivar - other.@ivar
+  #   end
+  # end
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:33]: ditto

```suggestion
  # class Foo
  #   def <=>(other : self)
  #     @ivar <=> other.@ivar
  #   end
  #
  #   def <=>(other : Foo)
  #     @ivar <=> other.@ivar
  #   end
  # end
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Lint/ReadInstanceVar:
+  #   Enabled: false
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:40]: These should list the default settings:

```suggestion
  #   Enabled: true
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  # ```
+  class ReadInstanceVar < Base
+    properties do
+      description "Disallows external reading of instance vars"
+    end
+
+    MSG = "Reading instance variables externally is not allowed."
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:47]: Let's stick to the already established stylistic convention:

```suggestion
    MSG = "Reading instance variables externally is not allowed"
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    def test(source)
+      InstanceVarReadVisitor.new(self, source)
+    end
+
+    def test(
+      source, node : Crystal::ReadInstanceVar,
+      allowed_names : Array(String)
+    ) : Nil
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:56]:
```suggestion
    def test(source, node : Crystal::ReadInstanceVar, allowed_names)
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          issue_for node, MSG, prefer_name_location: true
+        end
+      else
+        issue_for node, MSG, prefer_name_location: true
+      end
+    end
+
+    private class InstanceVarReadVisitor < AST::NodeVisitor
+      @class_name : String?
+      @param_names : Array(String) = [] of String
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:69]: Redundant type declaration

```suggestion
      @param_names = [] of String
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+          case restriction = arg.restriction
+          when Crystal::Path
+            if restriction.names[-1]? == @class_name
+              @param_names << arg.name
+            end
+          when Crystal::Self
+            @param_names << arg.name
+          end
+        end
+
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:91]:
```suggestion
```

File: src/ameba/rule/lint/read_instance_var.cr [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      @param_names : Array(String) = [] of String
+
+      def visit(node : Crystal::ClassDef) : Nil
+        prev_class = @class_name
+        @class_name = node.name.names[-1]?
+
+        node.body.accept(self)
+
+        @class_name = prev_class
+      end
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:78]:
```suggestion
      def visit(node : Crystal::ClassDef)
        prev_class = @class_name
        @class_name = node.name.names[-1]?
        begin
          node.body.accept(self)
        ensure
          @class_name = prev_class
        end
        true
      end
```

**Review by Sija — CHANGES_REQUESTED** [review]

**Comment by Sija** [comment]
Formatter on Crystal nightly is failing, btw.

File: src/ameba/rule/lint/read_instance_var.cr [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      description "Disallows external reading of instance vars"
+    end
+
+    MSG = "Reading instance variables externally is not allowed"
+
+    def test(source)
+      InstanceVarReadVisitor.new(self, source)
+    end
+
+    def test(source, node : Crystal::ReadInstanceVar, allowed_names)
```
Comment (Sija) [inline src/ameba/rule/lint/read_instance_var.cr:53]: Looks like the type restriction is actually needed here, my bad.

```suggestion
    def test(source, node : Crystal::ReadInstanceVar, allowed_names : Enumerable(String))
```

**Comment by straight-shoota** [comment]
As I mentioned before, it is impossible to have satisfactory type awareness without semantic analysis. 

You won't be able to determine the type of the target object with only syntactical information. The proposed implementation is very poor and limited. IIUC it only allows access to variables which are passed as parameters with explicit type restriction into a def. I'm not sure where the assumption comes from that valid access should only be allowed to parameters. It should be noted that a valid target object ought to be any kind of local variable, call, or even an instance variable read itself. But even there was justification for only considering parameters, there's still lots of issues. For example, the type of the variable at the time of the instance variable read does not need to be the same as in the method signature. Also, purely lexical comparison of paths may miss important details such as inheritance, aliases, generic arguments, ... I'm sure there are many more problems.

As a result, successful detection rate is quite miserable. IMO this is not even worth trying.
It should be better to invest the effort into figuring out how we could evolve ameba to make use of semantical information.

It's also worth noting that both human readers as well as the compiler (through semantic analysis) can easily understand the typing situation correctly. False positives on clear cases would be perceived as quite irritating issues from a linter.

**Comment by Sija** [comment]
@straight-shoota Good points, agreed.

**Comment by nobodywasishere** [comment]
I agree as well. I'll create an issue to start discussion of what integrating more semantic information into ameba would look like.

**Pull request closed without merging**